### ISSUE TITLE
EnableBroadcast property of the UdpClient class has no effect

### ISSUE THREAD

**Issue by gabschne** [dotnet/runtime#49711]
### Description

**Background:** I have a `UdpClient` and want to receive (directed) UDP broadcast packets with it. In addition to that, the packets should be only received from or on a specified network interface. For this I pass the IP of the targeted interface to the constructor of the `UdpClient` class using the `IPEndPoint` class. But when I tested it, I didn't receive any broadcast packets from it. I checked the docs and found the [`EnableBroadcast`](https://docs.microsoft.com/de-de/dotnet/api/system.net.sockets.udpclient.enablebroadcast?view=net-5.0) property.

According to the documentation, this property... 
> Gets or sets a Boolean value that specifies whether the UdpClient may send or receive broadcast packets.
[...]
`true` if the UdpClient allows broadcast packets; otherwise, `false`. The default is `false`.

So it seemed like this is the reason I didn't receive any broadcast packets. But in the next moment I thought "When the default is `false`, how am I able to broadcast with my test UDP sender **without** changing the property to `true` [and see it with Wireshark]?".

Setting the property in my client to `true` confirmed my worry: It still doesn't work.

---

**TL;DR:** The `EnableBroadcast` property of the `UdpClient` class has no effect. If it's set to `false` the client still sends and receives UDP broadcast packets. Furthermore, if the client is bound to a local IP address (or network interface) and set to `true`, it does not receive any either.

### Configuration

I'm using .NET 5 with the test UDP sender running on a Windows 10 x64 machine and the client running on a BeagleBone Green Wireless (ARM) with the official Debian Buster image (...so seems to not be a platform dependent problem).

### Regression?

Someone asked [a question on Stack Overflow in Nov. 2012 about this](https://stackoverflow.com/questions/13361779/udpclient-enablebroadcast-value-has-no-effect), so it doesn't seem to be a new problem.

### Other information
I didn't look into the code of the `EnableBroadcast` property, so I don't how it works (or should work). But I know that [there's a socket option for broadcasting](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socketoptionname?view=net-5.0#System_Net_Sockets_SocketOptionName_Broadcast). I tried this one, too. But using `Udp` as the `SocketOptionLevel` throws a `SocketException` with the code `10022` (invalid argument) on my sender (Windows) and code `95` (Operation not supported) on my client (Linux). If I set it to `Socket` no exception is thrown, however this also has no effect.

**Current workaround:** To know on which network interface the packet was received, I use the [`ReceiveMessageFrom`](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.receivemessagefrom?view=net-5.0) method of the `Socket` class. It has a `out IPPacketInformation` parameter which provides the index of the interface.

If this is not a bug and e.g. the property is only valid for some special use case(s), this may just be a documentation issue.

**Comment by karelz** [comment]
Triage: Can you please clarify which side (Windows or Linux) is failing?
Assumption is that on Linux sometimes only super-user have the ability to use these.

**Comment by gabschne** [comment]
> Triage: Can you please clarify which side (Windows or Linux) is failing?

As I wrote, I use the `UdpClient` (class) on a Windows machine to send (directed) UDP broadcast packages and on a BeagleBone (Linux) to receive them. 

If the `EnableBroadcast` property is set to `false`, the broadcast packets are still send from the windows machine. And if it's set to `true` while the `UdpClient` instance is bind to a network interface using it's IP, no packets are received on the BB/Linux side. But in the other hand, if the `UdpClient` is not bind to any interface and the property is set to `false`, it does receive them.

This means the answer to your question is "both", if I get it right.

> Assumption is that on Linux sometimes only super-user have the ability to use these.

The program has been executed with sudo.

**Comment by wfurt** [comment]
Could you share simple repro @gabschne ? On Linux this sets SO_BROADCAST
https://man7.org/linux/man-pages/man7/socket.7.html
The way how I read it it impacts only sending. So your receiving problem may live somewhere else.

**Comment by gabschne** [comment]
> The way how I read it it impacts only sending. So your receiving problem may live somewhere else.

Just checked this and can confirm. I tried to send broadcasts from the Beaglebone (Linux) and got a permission exception until I set the property to `true`.

But this and other platform dependent things are already information that might be worth stating in the documentation.

This also makes me think more and more that I have (at least) a thinking error in my code. I assumed that the `UdpClient` can receive broadcasts even if it's bind to a local endpoint, especially since there is a completely separate property for it.

But as I wrote in my initial description, some additional remarks in the docs might be helpful if it is not possible to bind the `UdpClient` to a local endpoint or network interface and still receive broadcast packets.

Here's a very quick repro code, of course you'll need to change the IP adresses to your own network and local endpoint:

### Sender
```
using System;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;

namespace UdpPacketSender
{
    class Program
    {
        private static async Task Main(string[] args)
        {
            const string text = "Hello there!";
            var data = Encoding.ASCII.GetBytes(text);
            using var udpClient = new UdpClient()
            {
                EnableBroadcast = true
            };

            Console.WriteLine($"EnableBroadcast property is set to {udpClient.EnableBroadcast}.");

            while (true)
            {
                await udpClient.SendAsync(data, data.Length, "192.168.178.255", 9009);
                await Task.Delay(500);
                Console.WriteLine("Sent.");
            }
        }
    }
}
```
### Receiver
```
using System;
using System.Net.Sockets;
using System.Threading.Tasks;

namespace UdpPacketReceiver
{
    class Program
    {
        static async Task Main(string[] args)
        {
            // If you bind a complete local endpoint, it doesn't receive any broadcasts.
            using var udpClient = new UdpClient(new IPEndPoint(IPAddress.Parse("192.168.178.89"), 9009))
            // By only setting/binding the port, it does receive broadcasts.
            // using var udpClient = new UdpClient(9009)
            {
                EnableBroadcast = true
            };

            Console.WriteLine($"EnableBroadcast property is set to {udpClient.EnableBroadcast}.");

            while (true)
            {
                var result = await udpClient.ReceiveAsync();
                Console.WriteLine($"Received data from {result.RemoteEndPoint.Address}");
            }
        }
    }
}
```

**Comment by wfurt** [comment]
The .NET is really just pass through exposing OS properties. It cannot really change underlying OS behavior @gabschne. 

The problem with receiver is that since you bind to `192.168.178.89` you tell OS to receive _only_ packets destined to that IP.... and  `192.168.178.255` will not match that. 

I used you Receiver  and when I pass in my `IPAddress.Parse("10.211.55.255"), 9009))` then I can receive broadcast no problem.
```
furt@ubu20:~$ echo foo | socat - udp-datagram:10.211.55.255:9009,broadcast


furt@ubu18:~/projects/broadcast$ ~/dotnet-5.x/dotnet run
EnableBroadcast property is set to True.
Received data from 10.211.55.16
```

You _could_ bind to wildcard IP and use bind to interface using trick described here: https://stackoverflow.com/questions/14478167/bind-socket-to-network-interface (needs root) 

There is no direct API in c# AFAIK so you would have to use [SetRawSocketOption](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.setrawsocketoption?view=net-5.0#System_Net_Sockets_Socket_SetRawSocketOption_System_Int32_System_Int32_System_ReadOnlySpan_System_Byte__) and little bit of platform code.

**Comment by gabschne** [comment]
> The problem with receiver is that since you bind to `192.168.178.89` you tell OS to receive _only_ packets destined to that IP.... and `192.168.178.255` will not match that.

Okay so I really just had some wrong assumptions of the local endpoint binding part and the `EnableBroadcast` property. 

> I used you Receiver and when I pass in my `IPAddress.Parse("10.211.55.255"), 9009))` then I can receive broadcast no problem.

Of course, that works for me, too. But then I receive the packets from all network interfaces. I have multiple interfaces in the same network and want to do some measurements, math and statistics only for a specific or given interface.

> You _could_ bind to wildcard IP and use bind to interface using trick described here: https://stackoverflow.com/questions/14478167/bind-socket-to-network-interface (needs root)
> 
> There is no direct API in c# AFAIK so you would have to use [SetRawSocketOption](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.setrawsocketoption?view=net-5.0#System_Net_Sockets_Socket_SetRawSocketOption_System_Int32_System_Int32_System_ReadOnlySpan_System_Byte__) and little bit of platform code.

Thank you for the tip! I'll try it out as soon as possible, but it seems like this meets my needs exactly. I really don't know why I hadn't found this in the research phase for this part of the project. But haven't had much to do with sockets at such a "low level", to be honest.

Regardless of whether the raw socket option will work for me or not, the `EnableBroadcast` property does seem to work, so sorry for all the noise. Feel free to close this issue if you don't see any further actions to be taken.
